WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,108 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	public class BranchAndBound : Selector
+	{
+		private Money _costOfHeader = Money.Satoshis(0);
+		private Money _costPerOutput = Money.Satoshis(0);
+		private int _bnbTryLimit = 5;
+		private Random _random = new();
+
+		private Money[]? UtxoSorted { get; set; }
+
+		public bool TryGetExactMatch(Money target, List<Money> availableCoins, out List<Money> selectedCoins)
+		{
+			selectedCoins = new List<Money>();
+			UtxoSorted = availableCoins.OrderByDescending(x => x.Satoshi).ToArray();
+			try
+			{
+				for (int i = 0; i < _bnbTryLimit; i++)
+				{
+					selectedCoins = RecursiveSearch(depth: 0, currentSelection: new List<Money>(), effValue: 0, target: target);
+					if (CalcEffectiveValue(selectedCoins) == target + _costOfHeader + _costPerOutput)
+					{
+						return true;
+					}
+				}
+
+				return false;
+			}
+			catch (Exception ex)
+			{
+				Logger.LogError("Couldn't find the right pair. " + ex);
+				return false;
+			}
+		}
+
+		private List<Money>? RecursiveSearch(int depth, List<Money> currentSelection, Money effValue, Money target)
+		{
+			var targetForMatch = target + _costOfHeader + _costPerOutput;
```

Comment:
This leads to 2 allocations. This costs probably milliseconds. 

Using `long` instead of `Money` would get rid of those allocations.

Reply:
```
private List<ulong>? RecursiveSearch(int depth, List<ulong> currentSelection, ulong effValue, ulong target)

```
Like this?

Reply:
I would have a look here: https://github.com/kiminuo/WalletWasabi/commit/178a1be98c359d06f254705588abef43900042b1. It's the iter 3 version mentioned here https://github.com/zkSNACKs/WalletWasabi/pull/6825#issuecomment-992237456.

Reply:
> ```
> private List<ulong>? RecursiveSearch(int depth, List<ulong> currentSelection, ulong effValue, ulong target)
> ```
> 
> Like this?

But if you want to take it slow, yes, you can do it like that too for now.